### ISSUE TITLE
[LOW][$500] LHN – Chat in LHN in not bold if receive a second IOU from new user

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#37315]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v1.4.44-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4343092
**Issue reported by:** Applause - Internal team

## Action Performed:
Preconditions:
2 accounts are needed for this test, 1 logged in the main testing account and the second in a secondary device or browser.
Both accounts should NOT have a existing conversation.
1. Go to https://staging.new.expensify.com/
2. Secondary device - Send a request money to the account in the main testing device. Make sure they don't have a existing conversation
3. Navigate to LHN
4. Verify the conversation is marked as unread
5. Navigate to the conversation
6. Navigate back to LHN (mobile) or to another conversation (Web/Desktop)
7. Verify the conversation is marked as read in LHN
8. Secondary device - Send another money request to the same account
9. Navigate to LHN

## Expected Result:
The conversation is marked as unread after a new request was made

## Actual Result:
The conversation is marked as read after a new request was made

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/93399543/d68a91cb-375e-4a4b-9781-b13b76fe966c

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014f4d3cd1fd4967db</li>
        <li>Upwork Job ID: 1762606281591746560</li>
        <li>Last Price Increase: 2024-03-04</li>
<li>Automatic offers: </li>
<ul>
<li>getusha | Reviewer | 0</li>
<li>FitseTLT | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @mallenexpensify</details>

**Comment by kbecciv** [comment]
@mallenexpensify I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by mallenexpensify** [comment]
@getusha do you think this can be external?
Adding to #vip-split project.

**Comment by getusha** [comment]
@mallenexpensify i think this can go external

cc @rlinoz looks like related to https://github.com/Expensify/App/pull/36276

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
LHN – Chat in LHN in not bold if receive a second IOU from new user

### What is the root cause of that problem?
We compare lastReadTime with lastVisibileActionCreated to determine `isUnread` and report action is created for the first request and for the second request only the request will be included inside the preview report action and no visible action is created and also `lastVisibleActionCreated` is not updated and hence the report is no indicated as unread
https://github.com/Expensify/App/blob/22cb01c968a53407f7a9d864bfba30f41366f6a0/src/libs/ReportUtils.ts#L3834-L3835
### What changes do you think we should make in order to solve the problem?

We should add another logic on `isUnread` function that will check for the latest request made by the other user and if the `created` time is greater than the `lastReadTime` the report will be considered as Unread
### What alternative solutions did you explore? (Optional)

**Comment by getusha** [comment]
>and also lastVisibleActionCreated is not updated

@FitseTLT do you mind adding a bit more details in your RCA?

**Comment by FitseTLT** [comment]
@getusha `lastVisibleActionCreated` is only updated when we create new visible report action and for the first time request a new `REPORTPREVIEW` type report action is created in the chat but for the second IOU request the user created in this issue the new request is included in the already existing `REPORTPREVIEW` type report action not new report action is created and hence `lastVisibleActionCreated` is not updated.

**Comment by getusha** [comment]
@FitseTLT's alternative [solution](https://github.com/Expensify/App/issues/37315#issuecomment-1976019278) sounds good to me.
🎀 👀 🎀 C+ Reviewed

**Comment by mallenexpensify** [comment]
@yuwenmemon 👀 on [the proposal above](https://github.com/Expensify/App/issues/37315#issuecomment-1976019278) when you have a min.

**Comment by mallenexpensify** [comment]
@FitseTLT when can we expect the PR?

**Comment by FitseTLT** [comment]
@yuwenmemon @getusha There are things we should discuss.
Let me first restate the problem to facilitate the discussion that follows below:
**Issue we are handling here:**
First I don't consider it as a bug but rather a `New Feature`. A new money request created will have a new report action created that represents it in the IOU Report and the IOU report is linked to some chat report. Now as the request is a visible report action  to the `IOU` report it will make it bold in LHN whenever a new request is created and when the user opens the IOU report new message marker will indicate the new requests created after the last read time. But for the chat report that is linked with the IOU report when the first request is created a Report Preview type report action will be created in the chat report but after that when other requests are created no new visible action is created but rather the no of requests that the report preview report action represents will increase.
********
Now with this background, what we are to achieve here is to indicate the user on the chat report whenever new requests are made that are relevant/linked to the chat report. 
********
On that I have two things to discuss:
1. @getusha selected my alternative solution which to calculate the `isUnread` by checking if there are new requests created after the chat report's `lastReadTime`. But to do that when we calculate `isUnread` of the chat report we need to be sure that all report actions linked to the report preview action (or the IOU report linked to the chat report) should exist in onyx. However, this is not always the case, the report actions will not be available until the user opens the relevant report (IOU report) for instance when a user first signin.
The correct way is my main proposal, which is same as we treated last user mentions by using `lastMentionedTime`. So we need to have a new prop on the chat report, like `lastRequestCreatedTime`, which will be populated by the BE whenever a new request (related to the chat report) is created and same as we handled mentions we will add another condition comparing it with `lastReadTime` to set the chat report unread here
https://github.com/Expensify/App/blob/6e42e55c2554c1dd5ab17e274f56ab5c4069eaf1/src/libs/ReportUtils.ts#L4012
2. After that from UX perspective we need to indicate the user that a new request is available similar to unread indicator we are currently using. The current unread indicator line we are currently using cannot be used for this case because we don't have new report action created in this issue's case instead we are trying to indicate the user that there is a new request created shown inside the report preview action. So we can may be display some msg above the composer or may be we can outline the report preview action with some different color to attract the user's attention. We can use any other method on this may be by consulting the design team 👍 

So @yuwenmemon let me know your thought as this need a BE change.

**Comment by mallenexpensify** [comment]
@yuwenmemon 👀 above, specifically wanting to know if this needs to be a backend change or not.  Thx

**Comment by yuwenmemon** [comment]
@thienlnam I feel like we were tackling something similar in our PRs with the `hasOutstandingChildRequest ` and `lastReadTime` for when a money request is transferred. Curious for your thoughts on the above...

**Comment by yuwenmemon** [comment]
@FitseTLT I'm sorry, can you simplify what you're trying to say? Perhaps: What exactly do you need from me in the backend?

**Comment by thienlnam** [comment]
> @thienlnam I feel like we were tackling something similar in our PRs with the hasOutstandingChildRequest  and lastReadTime for when a money request is transferred. Curious for your thoughts on the above...

This looks slightly different - I wonder if we should just update lastVisibleActionCreated when a new request is added on the same REPORTPREVIEW reportAction

**Comment by FitseTLT** [comment]
Sorry for the delay but will respond on Monday. Thx for the patience @yuwenmemon

**Comment by yuwenmemon** [comment]
Heads up I'm OOO next week so feel free to reassign to @thienlnam if you need any questions answered!

**Comment by yuwenmemon** [comment]
@FitseTLT Still waiting on you to address my follow-up question

**Comment by mallenexpensify** [comment]
Bumping back to `Daily` to ensure this is moving along.

**Comment by FitseTLT** [comment]
> @FitseTLT Still waiting on you to address my follow-up question

@yuwenmemon I have been investigating the feasibility of my main proposal that doesn't need a BE change so my plan is to create a draft PR in 3 working days from now and will discuss, test and proceed from there. Thnx.

**Comment by yuwenmemon** [comment]
Okay sounds good

**Comment by FitseTLT** [comment]
PR will be up in 2 days

**Comment by FitseTLT** [comment]
@yuwenmemon I tried on my main solution that doesn't need BE changes by checking on the IOU requests linked with the report preview of the chat report with this code
```
function doesChatHasNewRequest(report) {
    if (!isChatReport(report)) {
        return false;
    }

    const reportActions = allReportActions?.[`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${report?.reportID}`] ?? {};
    const sortedReportActions = ReportActionsUtils.getSortedReportActions(Object.values(reportActions), true);
    const latestReportPreviewAction = sortedReportActions.find((reportAction) => ReportActionsUtils.isReportPreviewAction(reportAction));

    if (!latestReportPreviewAction) {
        return false;
    }

    const iouReportID = ReportActionsUtils.getIOUReportIDFromReportActionPreview(latestReportPreviewAction);
    const iouReportActions = allReportActions?.[`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${iouReportID}`] ?? {};
    const latestIOURequestActionID = ReportActionsUtils.getMostRecentIOURequestActionID(Object.values(iouReportActions));

    if (!latestIOURequestActionID) {
        return false;
    }

    const latestIOURequestAction = iouReportActions[latestIOURequestActionID];
    const lastRequestCreatedTime = latestIOURequestAction.created;
    const iouReport = allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${iouReportID}`];

    return !!iouReport?.lastReadTime && iouReport.lastReadTime <= lastRequestCreatedTime && report.lastReadTime < lastRequestCreatedTime;
}
```

But the problem I have mentioned before has occurred, that is the `Unread` status was not reliable as it depends on the report action list of the iou report. For instance, what happened to me was, an unread chat report, without opening it, was unbolded when I search for reports (as I debugged it the report actions list was cleared from onyx at that momment) and then after a while it was back to its unread status when the report actions list was back to the onyx.

@yuwenmemon As I mentioned it before this issue is not a bug but rather a new feature request: which is, to set the chat report unread status when new requests are added to the report preview of that chat report. So if it is to be handled then the BE server should return `lastRequestCreatedTime` on the chat report and then we can calculate based on it as we did for last mention, here
https://github.com/Expensify/App/blob/34434a54704ff8e4df4929b9e51a39f29dfe131f/src/libs/ReportUtils.ts#L4639
 Or if it is not the right time to develop this new feature we might need to close this issue and develop this feature when the right time comes.

**Comment by yuwenmemon** [comment]
@rlinoz Does this make sense to you?

**Comment by mallenexpensify** [comment]
@rlinoz 👀 on [the above](https://github.com/Expensify/App/issues/37315#issuecomment-2080203078) plz

**Comment by mallenexpensify** [comment]
We have site issues and deploy freeze, I'll revisit next week. It sounds like we either want to close or get consensus that we want to update the feature.

**Comment by yuwenmemon** [comment]
Hmm, okay, I see. Yeah, I would tend to agree with you, @rlinoz. I don't think we want to necessarily create yet another new property on the report. I think we would rather just check the `lastModified` on the `REPORTPREVIEW` action.

**Comment by mallenexpensify** [comment]
Unsure of next best step here.  What cha think @rlinoz, @yuwenmemon @FitseTLT @getusha ?!?

**Comment by rlinoz** [comment]
I think we should check if it makes sense to use the `lastModified` property, if yes, then we need to check why it does not get updated and if we can update it, this will take an internal engineer though.

**Comment by mallenexpensify** [comment]
Will share next week in Slack if nobody comments before then.

**Comment by yuwenmemon** [comment]
This is marked as low priority so changing to a weekly. Will try and give it a look today though.

**Comment by mallenexpensify** [comment]
@yuwenmemon wanna take a look or should I try to rally someone in #engineering? 
It's #vip-split which isn't a priority now, so we could also put on hold or leave it a weekly and kick the can down the road til it's deemed a priority

**Comment by yuwenmemon** [comment]
Yeah, Monthly even.

**Comment by mallenexpensify** [comment]
I think I just experienced this. 
- Send request for $$$ to Deeter
- Navigate to a different chat
- Deeter send request to me for lunch money
- See this notification come in
<img width="408" alt="image" src="https://github.com/user-attachments/assets/9f22a38a-fa93-44d4-b955-dd7852c714b5">

- Observe LHN isn't bold
<img width="436" alt="image" src="https://github.com/user-attachments/assets/913ace6d-8fb7-46df-8789-2d3e1d6d7998">

Deeter then sent a chat and LHN **was** bold
<img width="406" alt="image" src="https://github.com/user-attachments/assets/0b736766-93ae-4fbf-ba28-008f53dd9c5d">

**Comment by yuwenmemon** [comment]
No update, bigger fish to fry as of now.

**Comment by mallenexpensify** [comment]
Throwing a `retest-weekly` on here since it's been open so long.

**Comment by FitseTLT** [comment]
not overdue

**Comment by mallenexpensify** [comment]
Waiting on retest

**Comment by mallenexpensify** [comment]
Still waiting on retest

**Comment by MelvinBot** [comment]
This has been labelled "Needs Reproduction". Follow the steps here: https://stackoverflowteams.com/c/expensify/questions/16989

**Comment by mallenexpensify** [comment]
Waiting on retest

**Comment by mallenexpensify** [comment]
Requesting retest in #qa    
https://expensify.slack.com/archives/C9YU7BX5M/p1738773971128549

**Comment by FitseTLT** [comment]
@mallenexpensify as I explained it before this is not a bug but more of a new feature.

As the main user A if you create an expense for user B for user B the chat report will be unread because the report preview corresponding to the expense you created is assumed as a new message for the chat report but now if from USER B if you open the chat so that it will be read and then if you navigate to other chat and if you User A again create an expense to user B chat that expense will be added to the report preview and new message/report action is not created on the main chat with user B so User B will not see bold in this case on the chat with the main user. If we want to give an indicator for the user that new expenses have been added on existing reports that is a new feature request and it needs BE changes but was deemed to be not the priority for now. 👍

**Comment by mallenexpensify** [comment]
Thanks @FitseTLT , one more thing.... 
In the examples above, it's User A and User B. And, in the title it says `Splits`.  Wouldn't this bug/new feature exist if I'm an employee (User A) and submit to my boss (User B)?  I _feel_ like this isn't just, only, for splitting.

**Comment by mallenexpensify** [comment]
@dannymcclain please hold on reviewing til we have a better idea of next steps.

**Comment by dannymcclain** [comment]
@mallenexpensify This feels more like a bug to me than a new feature? Shouldn't a new request always mark the chat as unread no matter what type of request?

**Comment by FitseTLT** [comment]
> [@mallenexpensify](https://github.com/mallenexpensify) This feels more like a bug to me than a new feature? Shouldn't a new request always mark the chat as unread no matter what type of request?

Nope @dannymcclain we only mark a chat as unread when new message/report action is created but we only create a new report preview report action for a new expense if there is no open expense/iou report in the chat. If there is a report that is not paid out then a new expense will be added to to that report and new message is only created for the iou/expense report not for the chat report. 

>I feel like this isn't just, only, for splitting.

@mallenexpensify Though the title says split the testing steps clearly show that it isn't so it is better to remove the Splits from the title 👍 This issue applies to both iou and expense reports.

**Comment by mallenexpensify** [comment]
Thanks @FitseTLT.  
@getusha and @FitseTLT , what do you propose is the best next step here? There are a lot of comments here. Do we want to update the logic so that we're marking a chat as unread if this action?
> new report preview report action for a new expense if there is no open expense/iou report in the chat.

**Comment by FitseTLT** [comment]
I think the best next step is to start a slack convo. I will do it on Monday if you are ok with it 👍

**Comment by mallenexpensify** [comment]
Please do @FitseTLT , post in #expensify-open-source and tag all assigned folks on this issue. Thx

**Comment by FitseTLT** [comment]
Created a convo https://expensify.slack.com/archives/C01GTK53T8Q/p1740078272966349

**Comment by mallenexpensify** [comment]
From convo in Slack, going to close. 

<img width="701" alt="Image" src="https://github.com/user-attachments/assets/64837cd1-e81b-41db-b886-612f87a197ea" />
